### PR TITLE
Create executor in exec_map only if requested.

### LINKED ISSUE #594: Segfaults due to executor generation in examples
Some of our examples use an executor map
```cpp
std::map<std::string, std::shared_ptr<gko::Executor>> exec_map{
    {"omp", omp},
    {"cuda", gko::CudaExecutor::create(0, omp, true)},
    {"hip", gko::HipExecutor::create(0, omp, true)},
    {"reference", gko::ReferenceExecutor::create()}};
```

Executing these tests with HIP and CUDA compiled causes a segmentation fault in my tests, probably due to the CUDA and HIP executor destroying the same device and/or library handles (hipSPARSE etc.)

This should be fixed by only creating the executor on demand, like in the benchmark examples (replacing std::shared_ptr<...> by std::function<std::shared_ptr<...>()> and using lambdas)

### PR DESCRIPTION
**Pull request by greole**
Fixes #594 Create executor only if requested using lambda functions inside the `exec_map`.

### CODE REVIEW

File: examples/custom-matrix-format/custom-matrix-format.cpp [hunk 1]
```diff
@@ -252,17 +252,25 @@ int main(int argc, char *argv[])
...
+                                                 true);
+             }},
+            {"reference", [] { return gko::ReferenceExecutor::create(); }}};
 
     // executor where Ginkgo will perform the computation
-    const auto exec = exec_map.at(executor_string);  // throws if not valid
+    const auto exec = exec_map.at(executor_string)();  // throws if not valid
     // executor used by the application
-    const auto app_exec = exec_map["omp"];
+    const auto app_exec = exec_map["omp"]();
```
Comment (upsj) [inline examples/custom-matrix-format/custom-matrix-format.cpp:273]:
```suggestion
    const auto app_exec = exec->get_master();
```

File: examples/nine-pt-stencil-solver/nine-pt-stencil-solver.cpp [hunk 2]
```diff
@@ -225,16 +225,25 @@ void solve_system(const std::string &executor_string,
...
+                                                 true);
+             }},
+            {"reference", [] { return gko::ReferenceExecutor::create(); }}};
+
     // executor where Ginkgo will perform the computation
-    const auto exec = exec_map.at(executor_string);  // throws if not valid
+    const auto exec = exec_map.at(executor_string)();  // throws if not valid
     // executor where the application initialized the data
-    const auto app_exec = exec_map["omp"];
+    const auto app_exec = exec_map["omp"]();
```
Comment (upsj) [inline examples/nine-pt-stencil-solver/nine-pt-stencil-solver.cpp:246]:
```suggestion
    const auto app_exec = exec->get_master();
```

File: examples/poisson-solver/poisson-solver.cpp [hunk 3]
```diff
@@ -134,17 +134,25 @@ int main(int argc, char *argv[])
...
+                                                 true);
+             }},
+            {"reference", [] { return gko::ReferenceExecutor::create(); }}};
 
     // executor where Ginkgo will perform the computation
-    const auto exec = exec_map.at(executor_string);  // throws if not valid
+    const auto exec = exec_map.at(executor_string)();  // throws if not valid
     // executor used by the application
-    const auto app_exec = exec_map["omp"];
+    const auto app_exec = exec_map["omp"]();
```
Comment (upsj) [inline examples/poisson-solver/poisson-solver.cpp:155]:
```suggestion
    const auto app_exec = exec->get_master();
```

File: examples/three-pt-stencil-solver/three-pt-stencil-solver.cpp [hunk 4]
```diff
@@ -160,16 +160,25 @@ void solve_system(const std::string &executor_string,
...
+                                                 true);
+             }},
+            {"reference", [] { return gko::ReferenceExecutor::create(); }}};
+
     // executor where Ginkgo will perform the computation
-    const auto exec = exec_map.at(executor_string);  // throws if not valid
+    const auto exec = exec_map.at(executor_string)();  // throws if not valid
     // executor where the application initialized the data
-    const auto app_exec = exec_map["omp"];
+    const auto app_exec = exec_map["omp"]();
```
Comment (upsj) [inline examples/three-pt-stencil-solver/three-pt-stencil-solver.cpp:181]:
```suggestion
    const auto app_exec = exec->get_master();
```

File: examples/twentyseven-pt-stencil-solver/twentyseven-pt-stencil-solver.cpp [hunk 5]
```diff
@@ -280,16 +280,25 @@ void solve_system(const std::string &executor_string,
...
+                                                 true);
+             }},
+            {"reference", [] { return gko::ReferenceExecutor::create(); }}};
+
     // executor where Ginkgo will perform the computation
-    const auto exec = exec_map.at(executor_string);  // throws if not valid
+    const auto exec = exec_map.at(executor_string)();  // throws if not valid
     // executor where the application initialized the data
-    const auto app_exec = exec_map["omp"];
+    const auto app_exec = exec_map["omp"]();
```
Comment (upsj) [inline examples/twentyseven-pt-stencil-solver/twentyseven-pt-stencil-solver.cpp:301]:
```suggestion
    const auto app_exec = exec->get_master();
```

File: contributors.txt [hunk 6]
```diff
@@ -17,3 +17,4 @@ Maier Matthias <matthias@43-1.org> Texas A&M University
 Nayak Pratik <pratik.nayak@kit.edu> Karlsruhe Institute of Technology
 Ribizel Tobias <mail@upsj.de> Karlsruhe Institute of Technology
 Tsai Yuhsiang <yhmtsai@gmail.com> National Taiwan University
+Gregor Olenik <go@hpsim.de> HPSim
```
Comment (yhmtsai) [inline contributors.txt:20]: we sort the contributor list by alphabet

Comment (thoasm) [reply contributors.txt:20]: Also, it is `last-name first-name`, so you would be below Pratik with
```
Olenik Gregor <go@hpsim.de> HPSim
```

**Comment by upsj** [comment]
Thanks for the contribution, you finished your first PR :) I guess we will merge @tcojean's PR first, but after that, you can merge yours (since you should already have developer permissions). Two important things about our workflow:

1. We try to keep our Git history clean and linear, thus after the previous PR is merged, you should run `git fetch && git rebase origin/develop` to rebase your changes onto the latest `develop` status (You can see how far ahead/behind develop you are when clicking on the branch name
Screenshot_2020-08-07 greole ginkgo

2. In our merge commits, we aim to give a meaningful name (not the default one) and description. The description should contain the URL of the PR in the end, so you can easily access it even when not coming from the Github web interface.
Screenshot from 2020-08-07 08-34-14

**Comment by tcojean** [comment]
Thanks for explaining the whole process @upsj. This time since we have a large backlog of PR to be merged I decided to go forward, but I advise @greole to take a look when he can on our merging policies. In the meantime, I will make sure to give you commit access.

**Pull request merged**